File: app/schemas/graph.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class GraphInfoResponse(BaseModel):
    total_nodes: int = Field(..., description="Total nodes in the graph")
    files: int = Field(..., description="Total file nodes")
    classes: int = Field(..., description="Total class nodes")
    functions: int = Field(..., description="Total function nodes")
    methods: int = Field(..., description="Total method nodes")
    modules: int = Field(..., description="Total module nodes")
    total_edges: int = Field(..., description="Total dependency edges")
    calls: int = Field(..., description="CALLS edges")
    imports: int = Field(..., description="IMPORTS edges")
    contains: int = Field(..., description="CONTAINS edges")
    defines: int = Field(..., description="DEFINES edges")
    belongs_to: int = Field(..., description="BELONGS_TO edges")


class CallerItem(BaseModel):
    id: Optional[str] = None
    name: str
    qualified_name: Optional[str] = None
    node_type: Optional[str] = None
    file_path: Optional[str] = None
    start_line: Optional[int] = None
    end_line: Optional[int] = None
    call_line: Optional[int] = None
    target_symbol: Optional[str] = None


class CallersResponse(BaseModel):
    symbol: str
    total_callers: int
    callers: List[CallerItem]


class CalleeItem(BaseModel):
    id: Optional[str] = None
    name: str
    qualified_name: Optional[str] = None
    node_type: Optional[str] = None
    file_path: Optional[str] = None
    start_line: Optional[int] = None
    end_line: Optional[int] = None
    call_line: Optional[int] = None
    caller_symbol: Optional[str] = None


class CalleesResponse(BaseModel):
    symbol: str
    total_callees: int
    callees: List[CalleeItem]


class DependencyItem(BaseModel):
    id: Optional[str] = None
    name: str
    node_type: Optional[str] = None
    file_path: Optional[str] = None
    start_line: Optional[int] = None
    end_line: Optional[int] = None
    call_line: Optional[int] = None
    depth: int
    call_path: Optional[str] = None
    called_by: Optional[str] = None


class DependenciesResponse(BaseModel):
    symbol: str
    depth: int
    total_dependencies: int
    dependencies: List[DependencyItem]


class DependentItem(BaseModel):
    id: Optional[str] = None
    name: str
    node_type: Optional[str] = None
    file_path: Optional[str] = None
    start_line: Optional[int] = None
    end_line: Optional[int] = None
    call_line: Optional[int] = None
    depth: int
    dependent_path: Optional[str] = None
    calls_target: Optional[str] = None


class DependentsResponse(BaseModel):
    symbol: str
    depth: int
    total_dependents: int
    dependents: List[DependentItem]


class ImpactItem(BaseModel):
    id: Optional[str] = None
    name: str
    node_type: Optional[str] = None
    file_path: Optional[str] = None
    start_line: Optional[int] = None
    end_line: Optional[int] = None
    call_line: Optional[int] = None
    calls_target: Optional[str] = None
    depth: Optional[int] = None


class ImpactResponse(BaseModel):
    symbol: str
    depth: int
    analysis_type: str = Field(default="STATIC DEPENDENCY IMPACT")
    total_impacted: int
    direct_callers: List[ImpactItem] = Field(default_factory=list)
    indirect_callers: List[ImpactItem] = Field(default_factory=list)
    direct_dependents: List[ImpactItem] = Field(default_factory=list)
    indirect_dependents: List[ImpactItem] = Field(default_factory=list)
    impacted_files: List[str] = Field(default_factory=list)

File: frontend/src/types/graph.ts
export interface GraphInfoResponse {
  total_nodes: number;
  total_edges: number;
  files: number;
  classes: number;
  functions: number;
  methods: number;
  modules?: number;
  calls: number;
  imports?: number;
  contains?: number;
  defines?: number;
  belongs_to?: number;
}

export interface CallerInfo {
  id?: string;
  name?: string;
  qualified_name?: string;
  caller_name?: string;
  caller_file?: string;
  caller_line?: number;
  file_path?: string;
  start_line?: number;
  end_line?: number;
  call_line?: number;
  call_type?: string;
  target_symbol?: string;
}

export interface GraphCallersResponse {
  symbol: string;
  total_callers: number;
  callers: CallerInfo[];
}

export interface CalleeInfo {
  id?: string;
  name?: string;
  qualified_name?: string;
  callee_name?: string;
  callee_file?: string;
  callee_line?: number;
  file_path?: string;
  start_line?: number;
  end_line?: number;
  call_line?: number;
  call_type?: string;
  caller_symbol?: string;
}

export interface GraphCalleesResponse {
  symbol: string;
  total_callees: number;
  callees: CalleeInfo[];
}

export interface DependencyItem {
  id?: string;
  name?: string;
  qualified_name?: string;
  target?: string;
  file?: string;
  file_path?: string;
  start_line?: number;
  end_line?: number;
  call_line?: number;
  depth: number;
  type?: string;
  caller?: string;
  call_path?: string;
}

export interface GraphDependenciesResponse {
  symbol: string;
  depth: number;
  total_dependencies: number;
  dependencies: DependencyItem[];
}

export interface DependentItem {
  id?: string;
  name?: string;
  source?: string;
  file?: string;
  file_path?: string;
  start_line?: number;
  end_line?: number;
  call_line?: number;
  depth: number;
  type?: string;
  calls_target?: string;
  dependent_path?: string;
}

export interface GraphDependentsResponse {
  symbol: string;
  depth: number;
  total_dependents: number;
  dependents: DependentItem[];
}

export interface GraphImpactItem {
  id?: string;
  name?: string;
  file_path?: string;
  node_type?: string;
  start_line?: number;
  end_line?: number;
  call_line?: number;
  calls_target?: string;
  depth?: number;
}

export interface GraphImpactResponse {
  symbol: string;
  depth: number;
  analysis_type: string;
  total_impacted: number;
  direct_callers: (GraphImpactItem | string)[];
  indirect_callers: (GraphImpactItem | string)[];
  direct_dependents?: (GraphImpactItem | string)[];
  indirect_dependents?: (GraphImpactItem | string)[];
  impacted_files: string[];
}

export interface GraphFileDependenciesResponse {
  file_path: string;
  imports_files?: string[];
  imports_modules?: string[];
  imports?: string[];
  imported_by: string[];
  defined_symbols?: any[];
  total_imports?: number;
  total_imported_by?: number;
}

